src/uwtools/rocoto.py

```diff
@@ -25,20 +33,113 @@ def _add_jobname(tree: dict) -> None:
             _add_jobname(subtree)
 
 
-# Public functions
-def write_rocoto_xml(input_yaml: str, input_template: str, rendered_output: str) -> None:
+def _add_tasks(
+    input_yaml: OptionalPath = None,
+) -> None:
     """
-    Main entry point.
+    Define "task" elements in the given config tree and request a "jobname" for each.
 
     :param input_yaml: Path to YAML input file.
-    :param input_template: Path to input template file.
-    :param rendered_output: Path to write rendered XML file.
     """
     values = YAMLConfig(input_yaml)
     tasks = values["tasks"]
     if isinstance(tasks, dict):
         _add_jobname(tasks)
 
+
+# Public functions
+def realize_rocoto_xml(
+    input_yaml: OptionalPath = None,
+    rendered_output: OptionalPath = None,
+) -> bool:  # pragma: no cover
+    """
+    Realize the given YAML file to XML, using fixed Rocoto RelaxNG schema templates. Validate both
+    the YAML and the XML. External functions should call this function.
+
+    :param input_yaml: Path to YAML input file.
+    :param input_template: Path to input template file.
+    :param rendered_output: Path to write rendered XML file.
+    :param schema_file: Path to schema file.
+    """
+
+    _add_tasks(input_yaml)
+
+    rocoto_schema = _rocoto_schema()
+    rocoto_template = _rocoto_template()
+
+    # Validate the YAML.
+    if uwtools.config.validator.validate_yaml(config_file=input_yaml, schema_file=rocoto_schema):
+        # Render the template.
+        write_rocoto_xml(
+            input_yaml=input_yaml,
+            input_template=rocoto_template,
+            rendered_output=rendered_output,
+        )
+        # Validate the XML.
+        if validate_rocoto_xml(input_xml=rendered_output, schema_file=rocoto_template):
+            # If no issues were detected, report success.
+            return True
+        logging.error("Rocoto validation errors identified in %s", rendered_output)
+        return False
+    logging.error("YAML validation errors identified in %s", input_yaml)
+    return False
+
+
+def _rocoto_template() -> str:
+    """
+    The path to the file containing the template to validate the config file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return (path / "rocoto.jinja2").as_posix()
+
+
+def _rocoto_schema() -> str:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return (path / "rocoto.jsonschema").as_posix()
+
+
+def validate_rocoto_xml(input_xml: OptionalPath = None, schema_file: OptionalPath = None) -> bool:
+    """
+    Given a rendered XML file, validate it against the Rocoto schema.
+
+    :param input_XML: Path to rendered XML file.
+    :param schema_file: Path to schema file.
+    """
+
+    # Validate the XML.
+    with readable(schema_file) as f:
+        schema = etree.RelaxNG(etree.parse(f))
```

Comment:
I'm not sure that `readable()` is appropriate here: It is if we want to give users an option to pipe in a schema via `stdin`, but I think from other comments/discussion we were thinking that *our* schema, the one in `resources/` is the one to use, and that it might not make sense to give users the ability to override that -- which makes sense to me. If that's what we collectively think, than should we have both of these functions that the CLI is dispatching to simply obtain the correct file via `resources/` and not present the user with a choice they shouldn't be making? Apologies if I missed something in discussion on this point, though...

Reply:
Yes, I was confused about what we were validating here -- this is the XML output, not the YAML input. But I think the crucial question remains: Do we want to present users with an option to override the official RelaxNG schema that we took from the Rocoto repo? I can't immediately see why, so I'm on team YAGNI on this one.

Reply:
Yep, removing all the schema options!

Reply:
I'm getting confused about this one -- can we discuss briefly post-standup today?

Reply:
I'll summarize my misgivings (misunderstandings?) here as a prelude to our upcoming meeting:

1. In other modes, we permit provision of the "primary" input on `stdin` as an alternative to providing an on-disk file. For example, with `uw template render`, the Jinja2 template to be rendered can be specified via the _optional_ `--input-file` argument, or it can be piped in on `stdin`. The latter means someone could do something like `curl -s https://foo.com/path/to/template.jinja2 | uw template render` and render a template pulled from the internet. Why? Who knows, but it's the "Unix way" and can be handy in lots of circumstances. It could also be `cat`ed or produced however, and piped in. My (possibly wrong) understanding of `uw rocoto validate` is that we have a would-be Rocoto XML document that we want to validate, so to me this would be the "primary" input of this mode, so it ought to -- for consistency with other modes -- be an optional argument, and taken from `stdin` if not specified, which is what `readable()` provides: Either a readable handle to an opened file, if a filename was provided, or a readable handle to `stdin` if not. So I kind of expected to see `--input-file` optional, and its value handled by `readable()` for consistency with other modes.
2. I think `_rocoto_template()` is probably misnamed: The file it returns a path to is not a template, it's a schema. Of course, the name `_rocoto_schema()` is already taken, so might we have something like `_rocoto_schema()` -> `_rocoto_schema_yaml()` and `_rocoto_template()` -> `_rocoto_schema_xml()`? Is that what the files at the two returned paths are, namely the schema for a Rocoto workflow document in YAML form, and a Rocoto workflow document in XML form?
3. Finally, I wouldn't expect to see `readable()` used to get a handle to the Rocoto XML schema file. I mean, it _works_, but the use of that function -- if you know what that function does -- gives the appearance that that file might be provided on `stdin`, which is not the case with the schema file. Per point 1 above, I'd hope to see the XML document optionally available on `stdin`, so would expect to see `readable()` used to get a readable handle to it; but not to the schema file. The schema file is simply an on-disk file, so I'd just expect to see a `with open(...) ...` block on that one. Again, `readable()` works because we'd never _not_ provide a real filename, but I think `readable()` sends the wrong message. I _think_ that if you look for the existing `with readable()` statements in the codebase, you'll find that they're all tied to inputs that might be unspecified, and where `stdin` will be used instead; while there are other `with open()` calls for cases where the input will always be a file. See e.g. `uwtools.config.validator.validate_yaml()`, where `with open()` is used to read in the schema file. I think adhering to this convention will remind us which files may be streams, and which must always be files.

Reply:
Looking through, I may have to adjust functions to `with readable()` all the way up to `config.YAMLConfig()` but we'll definitely talk about it as a group

Reply:
Feedback from the offline discussion:
1) We will be moving forward with converting to streams, but that will move to another PR to cover a broader scope
2&3) Agreed and added, as above
also, reiterating the replacement of `str()` with `Path()` which is now a requirement of this PR and is in progress
Thanks to @maddenp-noaa, @christinaholtNOAA and @NaureenBharwaniNOAA for their time today!

Reply:
Hmm, `j2template()` and `config()` are fixed, I think the only remnant `str` paths may be in `forecast()` now and some redundant stragglers in tests

Reply:
This one looks good to me now -- please feel free to resolve the conversation if you have no misgivings.